### DISCUSSION TITLE
Read `umbraco-package.json` from child directories.

### DISCUSSION THREAD

**Discussion by markadrake** [umbraco/Umbraco-CMS#16171] [Features and ideas]
I'm exploring the Umbraco 14 RC and writing an RCL library.

I noticed if my `app-plugins` directory has this composition, that the `umbraco-package.json` is ignored.

- `app_plugins`
   - `package`
      - `module`
         - `umbraco-package.json` _ignored_ 
      - `umbraco-package.json`

[This may be the code responsible.](https://github.com/umbraco/Umbraco-CMS/blob/d4188022af2a0c7d8ebc6214ff4ed1d29f7865ed/src/Umbraco.Infrastructure/Manifest/PackageManifestReader.cs#L43)

**Comment by iOvergaard** [comment]
Hi, this was my decision, so going to address it here 🙋‍♂️ 

We noticed this was a thing in Umbraco 13 and earlier -- it would pick up the then `package.manifest` file from an unlimited amount of subfolders. Then people started putting their node_modules folder inside app_plugins and boot-up performance went down the drain (we cache the files, but still, what if one of your node_modules also publishes such a file?). 

We decided it would make sense if the new `umbraco-package.json` file would only be picked up from either the root of the app_plugins folder or one level down (i.e. in a package folder).

We could potentially introduce another level, so this would work:

```
App_Plugins/
  - Package/
    - Module/
      - umbraco-package.json
```

Could you perhaps share your thoughts towards the use-case of this, @markadrake?

**Reply by markadrake** [reply]
@iOvergaard Wow, the things you and your team have to deal with! I can't believe people are publishing the `node_modules` directory here. I would have never thought of that. 😭

Thank you for confirming that this method did work in previous versions! I'm new to writing RCL packages and thought it was something I may have done to myself.

I can compromise–knowing now of the performance issue you described. I recently published some of the packages I was working on and the route I took was to merge the package and module name so there's only one folder now.

I like to "namespace" my App_Plugins directory:

```
App_Plugins/
     - Humble.Umbraco/
          - CustomBackofficeIcons/
          - SVGHelpers/
          - ... (etc)
```

Do you think enough people are going to try placing their `umbraco-package.json` in child folders that [an update to this page would be helpful](https://docs.umbraco.com/umbraco-cms/v/14.latest-rc/extending-backoffice/development-flow/package-manifest)?

Thanks again for your time and consideration @iOvergaard!

**Reply by skttl** [reply]
Another possible solution could be having an index file in the root folder, containing paths to all umbraco-package.json files found further down in the tree.

**Reply by iOvergaard** [reply]
@markadrake [The article you mentioned](https://docs.umbraco.com/umbraco-cms/extending/package-manifest) has just today been updated with the new format including a sentence about the file having to be a maximum of two levels down.

**Reply by iOvergaard** [reply]
I also included a paragraph on writing your own IPackageManifestReader, if you want to customize where the files are found. Hope that makes sense! 👍
